geode-core/src/main/java/org/apache/geode/internal/cache/EntryEventImpl.java

```diff
@@ -2124,16 +2158,24 @@ public String toString() {
     if (Boolean.getBoolean("gemfire.insecure-logvalues")) {
       buf.append(";oldValue=");
       try {
-        synchronized (this.offHeapLock) {
-          ArrayUtils.objectStringNonRecursive(basicGetOldValue(), buf);
+        if (mayHaveOffHeapReferences()) {
+          synchronized (this.offHeapLock) {
+            ArrayUtils.objectStringNonRecursive(basicGetOldValue(), buf);
+          }
+        } else {
+          buf.append("OFFHEAP_VALUE_FREED");
```

Comment:
This is not correct. If it is not off-heap then you just want to do this:
ArrayUtils.objectStringNonRecursive(basicGetOldValue(), buf);
Also you should move the try/catch inside the synchronized(this.offHeapLock) block.
It is only when this block converts the value to a string that it can throw IllegalStateException because the off heap value has been freed.